Support for MS database tools and assemblies

JeanRev opened the pull request:
This pull request targets #1095. Some of the changes made are however going a bit beyond the scope of this issue.

I divided the resulting branch into multiple commits to keep a better overview.

The first two commits (4941aab and e66fc54) extends the current unit tests to include some specific sub-type of database objects, that are handled explicitly in the new clean-up logic implemented in 49a3c99:
- 4941aab: test inlined and table-valued function.
- e66fc54: test functions and procedure based on CLR assemblies.

All the revised tests are passing with no modification on the productive code.

49a3c99: add the effective handling of objects generated from Microsoft database tools, along with some specific unit tests.

0cb8f35: add support for cleaning CLR assemblies as well as an additional SQL server object type: user-defined aggregates.

JeanRev commented on flyway-core/src/test/resources/migration/dbsupport/sqlserver/sql/assembly/FlywayTestAssembly.cs:
```diff
@@ -0,0 +1,173 @@
+/**
```
The CLR assembly source code. Must be manually compiled and included in the test script as an hexadecimal string (see generate_assembly.txt).

JeanRev commented on flyway-core/src/test/resources/migration/dbsupport/sqlserver/sql/assembly/V1__assembly.sql:
```diff
@@ -0,0 +1,59 @@
+
+CREATE ASSEMBLY flyway_test_assembly
+FROM 0x4D5A90000300000004000000FFFF0000B800000000000000400000000000000000000000000000000000000000000000000000000000000000000000800000000E1FBA0E00B409CD21B8014CCD21546869732070726F6772616D2063616E6E6F742062652072756E20696E20444F53206D6F64652E0D0D0A2400000000000000504500004C01030077FD00570000000000000000E00002210B0108000010000000060000000000009E2E0000002000000040000000004000002000000002000004000000 ...
+WITH PERMISSION_SET = SAFE /* SAFE means that the assembly will not be allowed to access external resources such as
```
This huge hexadecimal value is the assembly DLL file content.

They are some obvious safety concerns when integrating binary content in a code repo. However the involved risk is here relatively low since it is only for test purpose, and the configured safety level should forbid to do anything nasty outside of the Database...

JeanRev commented on flyway-core/src/test/resources/migration/dbsupport/sqlserver/sql/assembly/generate_assembly.txt:
```diff
@@ -0,0 +1,10 @@
+The test assembly (.cs file) can be compiled against the .Net 2.x framework or greater.
```
Guidelines to generate and integrate the assembly...

JeanRev commented on flyway-core/src/test/java/org/flywaydb/core/internal/dbsupport/sqlserver/SQLServerMigrationTestCase.java:
```diff
@@ -155,6 +157,65 @@ public void type() throws Exception {
+    @Test
+    public void assembly() throws Exception {
+
+        CallableStatement stmt = jdbcTemplate.getConnection().prepareCall("EXEC sp_configure 'clr enabled', 1; RECONFIGURE;");
```
Required configuration statement in order to run CLR code.

JeanRev commented on flyway-core/src/test/java/org/flywaydb/core/internal/dbsupport/sqlserver/SQLServerMigrationTestCase.java:
```diff
@@ -155,6 +157,65 @@ public void type() throws Exception {
+            flyway.migrate();
+        } finally {
+            try {
+                jdbcTemplate.getConnection().prepareCall("EXEC sp_configure 'clr enabled', 0; RECONFIGURE;");
```
Disable CLR execution when done.

JeanRev commented on flyway-core/src/main/java/org/flywaydb/core/internal/dbsupport/sqlserver/SQLServerSchema.java:
```diff
@@ -112,25 +214,80 @@ protected void doClean() throws SQLException {
+     * @return the found objects
+     * @throws SQLException when the retrieval failed
+     */
+    private List<DBObject> queryDBObjects(ObjectType... types) throws SQLException {
```
Instead of relying on several tables of the information_schema, the proposed code makes only use of the sys.objects table whenever possible. This make joins based on object ids (required to query extended property where the Microsoft tools flags are stored) very straightforward.

However, this imply that we also must select a bit more precisely the objects we are interested only (not only function and procedure, but low-level types such as "SQL function", "CLR function", "SQL Procedure", "CLR procedure, etc."). Thus, the additional testing required and implemented in the two first commit of this PR.

JeanRev commented on flyway-core/src/test/java/org/flywaydb/core/internal/dbsupport/sqlserver/SQLServerMigrationTestCase.java:
```diff
@@ -327,6 +329,52 @@ public void dmlErrorsCorrectlyDetected() throws Exception {
   }
 
+    @Test
+    public void msDBToolsIgnoredForEmpty() throws Exception {
```
Testing: create the MS tools assets do ensure that they are ignored on doEmpty and doClean.

JeanRev commented:
One possible improvement I'm seeing, would be to divide the new tests assets into separate unit test, in order to test more finely the behavior of doClean/doEmpty...

axelfontaine commented on flyway-core/src/test/resources/migration/dbsupport/sqlserver/createMSDBTools.sql:
```diff
@@ -0,0 +1,556 @@
+/*
+Create Visual Studio tools in schema. Profiled from SQL Server Management studio 2014. For testing purpose only.
+ */
+IF OBJECT_ID(N'dbo.sp_upgraddiagrams') IS NULL and IS_MEMBER('db_owner') = 1
```
What is the license of these stored procedures? It seems you copied them one to one from somewhere. Can they freely be cloned? What is the original source?

JeanRev replied:
They were profiled from a running MS SQL Studio instance (that is what I intended to say with the comment at the top of the file).

If you think that it could be a problem, I can also write a test script from scratch...

alkamo replied:
They are provided by Microsoft and are built automatically when creating diagrams using MS SQL Studio.

IP concerns aside, this is potentially a versioning nightmare (Which version of SQL Server were these extracted from? Which versions are they compatible with?).

JeanRev replied:
Ok, new test scripts were committed in fc4aff1.

axelfontaine commented:
Thanks for the update and the removal of the sources. Looks good now. Could you make this mergeable again?

JeanRev commented:
@axelfontaine of course! I'll need some time to restore my test infrastructure and to review the code once more before merging (it's been awhile).

JeanRev commented:
Ok, I have rebased the PR on master (2d472b8) and removed the old test script from history.

In addition, two conflicts were encountered in the SQLServerSchema class, reflecting two upstream changes:
- 726d0d4: Uppercasing of the "PROCEDURE" and "FUNCTION" routine type value. This commit was ignored (do not apply to the new code).
- d46a406: queried objects must be sorted using their creation date, was added in this PR as commit c872439 (not doing so effectively trigger a test failure related to #1436).

Otherwise, this PR is unchanged.

axelfontaine commented:
Thanks! Looks great now! 👍 

This will be part of the next Flyway release.

JeanRev commented:
Great :-) Looking forward to it!

axelfontaine commented:
Merged. Thanks!
